File: providers/tristar/src/main/kotlin/gdansk/provider.kt
package app.wheretopark.providers.tristar.gdansk

import app.wheretopark.providers.tristar.Provider
import app.wheretopark.shared.*
import com.charleskorn.kaml.Yaml
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.time.DayOfWeek

const val METADATA_URL = "https://ckan.multimediagdansk.pl/dataset/cb1e2708-aec1-4b21-9c8c-db2626ae31a6/resource/d361dff3-202b-402d-92a5-445d8ba6fd7f/download/parking-lots.json"
const val STATE_URL = "https://ckan2.multimediagdansk.pl/parkingLots"

@Serializable
data class ParkingLotConfiguration(
    val resources: List<ParkingLotResource>,
    @SerialName("total-spots")
    val totalSpots: UInt,
    val features: List<ParkingLotFeature>,
    val rules: List<ParkingLotRule>
)

@Serializable
data class Configuration(
    @SerialName("parking-lots")
    val parkingLots: Map<String, ParkingLotConfiguration>,
)

class TristarGdanskProvider: Provider() {
    private val client = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
                ignoreUnknownKeys = true
            })
        }
    }
    // Mapping of vendor and storekeeper ID's
    private val ids = mutableMapOf<String, ParkingLotID>()

    private val configuration = Yaml().decodeFromString<Configuration>(this.javaClass.getResource("/configuration-gdansk.yaml")!!.readText())

    override suspend fun metadatas(): Map<ParkingLotID, ParkingLotMetadata> {
        val vendorMetadatas = client.get(METADATA_URL).body<MetadataResponse>()
        return vendorMetadatas.parkingLots.map { it ->
            val configuration = configuration.parkingLots[it.id] ?: return@map null
            val id = it.location.hash()
            ids[it.id] = id
            val metadata = ParkingLotMetadata(
                name = it.name,
                address = it.address,
                location = it.location,
                resources = configuration.resources,
                totalSpots = mapOf(
                    Pair(ParkingSpotType.CAR, configuration.totalSpots)
                ),
                features = configuration.features,
                currency = "PLN",
                rules = configuration.rules,
            )
            id to metadata
        }
        .filterNotNull()
        .toMap()
    }

    override suspend fun states(): Map<ParkingLotID, ParkingLotState> {
        val vendorStates = client.get(STATE_URL).body<StateResponse>()
        return vendorStates.parkingLots.map {
            val id = ids[it.id] ?: return@map null
            val state = ParkingLotState(
                lastUpdated = it.lastUpdate,
                availableSpots = mapOf(
                    Pair(ParkingSpotType.CAR, it.availableSpots)
                ),
            )
            id to state
        }
        .filterNotNull()
        .toMap()
    }
}

File: providers/tristar/gdansk/provider.go
package gdansk

import (
	"errors"
	wheretopark "wheretopark/go"

	geojson "github.com/paulmach/go.geojson"
)

type Provider struct {
	configuration Configuration
	mapping       map[string]wheretopark.ID
}

func (p Provider) GetMetadata() (map[wheretopark.ID]wheretopark.Metadata, error) {
	vendorMetadata, err := GetMetadata()
	if err != nil {
		return nil, err
	}
	metadatas := make(map[wheretopark.ID]wheretopark.Metadata)
	for _, vendor := range vendorMetadata.ParkingLots {
		configuration := p.configuration.ParkingLots[vendor.ID]
		id := wheretopark.CoordinateToID(vendor.Location.Latitude, vendor.Location.Longitude)
		metadata := wheretopark.Metadata{
			Name:           vendor.Name,
			Address:        vendor.Address,
			Location:       *geojson.NewPointFeature([]float64{vendor.Location.Longitude, vendor.Location.Latitude}),
			Resources:      configuration.Resources,
			TotalSpots:     configuration.TotalSpots,
			MaxWidth:       configuration.MaxWidth,
			MaxHeight:      configuration.MaxHeight,
			Features:       configuration.Features,
			PaymentMethods: configuration.PaymentMethods,
			Comment:        configuration.Comment,
			Currency:       "PLN",
			Timezone:       "Europe/Warsaw",
			Rules:          configuration.Rules,
		}
		metadatas[id] = metadata
		p.mapping[vendor.ID] = id
	}
	return metadatas, nil
}

func (p Provider) GetState() (map[wheretopark.ID]wheretopark.State, error) {
	vendorState, err := GetState()
	if err != nil {
		return nil, err
	}
	states := make(map[wheretopark.ID]wheretopark.State)
	for _, vendor := range vendorState.ParkingLots {
		id, exists := p.mapping[vendor.ID]
		if !exists {
			return nil, errors.New("no mapping for " + vendor.ID)
		}

		state := wheretopark.State{
			LastUpdated: vendor.LastUpdate,
			AvailableSpots: map[string]uint{
				"CAR": vendor.AvailableSpots,
			},
		}
		states[id] = state
	}
	return states, nil
}

func NewProvider(configurationPath string) (*Provider, error) {
	configuration, err := LoadConfiguration(configurationPath)
	if err != nil {
		return nil, err
	}
	return &Provider{
		configuration: *configuration,
		mapping:       make(map[string]wheretopark.ID),
	}, nil

}
